src/middlewares.ts

```diff
@@ -15,37 +15,85 @@ import {
 	touchCommit,
 } from './api';
 
-export async function determineCommitHash(req: any, res: any, next: any) {
-	const isHashInSession = !!req.session.commitHash;
-	const isHashSpecified = req.query && (req.query.hash || req.query.branch);
-	let commitHash;
+function assembleSubdomainUrlForHash(req: any, commitHash: string) {
+    const protocol = (req.secure ? 'https' : 'http');
 
-	if (isHashInSession && !isHashSpecified) {
-		next();
-		return;
-	}
+    return protocol + '://hash-' + commitHash + '.' + stripCommitHashSubdomainFromHost(req.headers.host);
+}
 
-	if (!isHashSpecified) {
-		res.send('Please specify a branch to load');
-		return;
-	}
+function stripCommitHashSubdomainFromHost(host: string) {
+    let segments = host.split('.'),
+        commitHashIndex = null;
+
+    for (let i = 0; i < segments.length; i += 1) {
+        if (0 === segments[i].indexOf('hash-')) {
+            commitHashIndex = i;
+            break;
+        }
+    }
+
+    if (null === commitHashIndex) {
+        return host;
+    }
+
+    return segments.slice(commitHashIndex + 1).join('.');
+}
```

Comment:
it would be helpful if we had a comment with some example input and output here as I'm finding the logic hard to parse in my head. what are we trying to accomplish here? are we trying to take `hash-12345.calypso.live/me` and produce `calypso.live/me` from it?

alternatively some unit tests could clarify the expected usage


we might be able to simplify the logic by removing some of the procedural steps and focus more on the semantic transformation?
```js
function withoutHashSubdomain( host ) {
	return host.replace( /(^|.*?\.)hash-[a-f0-9]+\./, '' );
}
```

I payed around with this in a jsbin - http://jsbin.com/cemogujami/1/edit?js,console - but I was unable to infer what the intended transformation was so the code could be wrong - you can verify by looking at the example input strings and see if the outputs produced by the `regexp` function are the correct transformed host strings.